### ISSUE TITLE
Change Request: Add Markdown, JSON, and CSS language plugins to `eslint-config-eslint`

### ISSUE THREAD

**Issue by lumirlumir** [eslint/eslint#20585]
### ESLint version

HEAD

### What problem do you want to solve?

This issue originates from an older comment: https://github.com/eslint/eslint.org/pull/840#pullrequestreview-3500977455.

Given the current PRs affecting `eslint-config-eslint` (https://github.com/eslint/eslint/pull/20576, #20584), it seems ready for the next breaking change.

JSON v1.0.0 has been released, CSS v1.0.0 is imminent and only awaiting release, and the Markdown v8.0.0–related PRs are open and ready for review. Therefore, I'd like to propose adding the Markdown, JSON, and CSS plugins to `eslint-config-eslint`, when the next versions are all ready.

### What do you think is the correct solution?

I think adding new subpaths such as `eslint-config-eslint/markdown`, `eslint-config-eslint/json`, and `eslint-config-eslint/css` to follow existing convention would be one option. Other suggestions are welcome.

### Participation

- [x] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

I can apply this change to all repositories under ESLint and will track progress if there's consensus on this change request.

**Comment by fasttime** [comment]
I think we're not currently linting any CSS files. Do you have a new use case in mind? As for adding JSON and Markdown, I'd be in favor. It would help the discussion though if you could outline which plugins, rules, and other settings you're thinking of including in those configs.

**Comment by nzakas** [comment]
`eslint-config-eslint` is meant to be the core linting configuration for all ESLint projects, so we should make sure whatever is in there applies to all ESLint projects. I'd be in favor of adding JSON and Markdown, and to @fasttime's point, we'd need to decide on which rules to enable beyond the recommended set (although we could start with recommended and add more later).

**Comment by lumirlumir** [comment]
Sorry for the delay. I’ve gone through the rules, and below are the rulesets I believe would be best to apply.

(Since Markdown has many more rulesets, I first looked into the CSS and JSON rulesets and configurations. I'll update the Markdown rulesets soon.)

## CSS

> I think we're not currently linting any CSS files. Do you have a new use case in mind?

Only the `code-explorer` repository is currently using a vanilla CSS file alongside Tailwind CSS, so I think CSS rules could be applied there. Since it's just one repository, I don't have a strong opinion. But, I think it's probably worth adding so we can dogfood it.

- https://github.com/eslint/code-explorer/blob/main/src/app.css
- https://github.com/eslint/code-explorer/blob/main/src/codemirror-themes.css

## JSON

### Rules

- ✅ no-duplicate-keys: 
    - Since this is part of the recommended ruleset, I'm in favor of adding it. 
    - This rule has no options.

- ✅ no-empty-keys:
    - Since this is part of the recommended ruleset, I'm in favor of adding it. 
    - This rule has no options.

- ✅ no-unnormalized-keys:
    - Since this is part of the recommended ruleset, I'm in favor of adding it. 
    - This rule has a single option, and I think the default `form: "NFC"` makes sense here.

- ✅ no-unsafe-values:
    - Since this is part of the recommended ruleset, I'm in favor of adding it. 
    - This rule has no options.

- ⚠️ sort-keys:
    - This rule is not part of recommended ruleset.
    - I think using the `sort-keys` rule only for `package.json` would be useful, since all fields can be sorted and it would ensure consistency across projects. I'd like to raise this for further discussion.

- ✅ top-level-interop:
    - This rule is not part of recommended ruleset.
    - This rule has no options.
    - This rule requires the top-level JSON value to be an array or object. As far as I recall, there aren’t any JSON files that use primitive top-level values, so I think it’s a good rule to enable.


### Configurations

The draft configuration is as follows. Further discussion is welcome.


```js
const json = require('@eslint/json').default;

// ...

const jsonConfigs = [
  {
    plugins: {
      json,
    },
  },

  // JSON
  {
    name: 'eslint-config-eslint/json',
    files: ['**/*.json', ".c8rc"],
    ignores: [
      '**/package-lock.json', // Ignored because lock files are usually auto generated.
      '**/.vscode/*.json', // VS Code settings files can use JSONC features.
      '**/jsconfig.json', // I'm not sure if `jsconfig.json` is useful here, but it's used in the same way as `tsconfig.json`. Further discussion is welcome.
      '**/jsconfig.*.json', // I'm also not sure about this. Further discussion is welcome.
      '**/tsconfig.json', // `tsconfig.json` can use JSONC features.
      '**/tsconfig.*.json', // There are `tsconfig.base.json` and `tsconfig.dist.json` files in `rewrite` repository.
    ],
    language: 'json/json',
    rules: {
      'json/no-duplicate-keys': 'error',
      'json/no-empty-keys': 'error',
      'json/no-unnormalized-keys': 'error',
      'json/no-unsafe-values': 'error',
      'json/top-level-interop': 'error',
    },
  },

  // JSONC
  {
    name: 'eslint-config-eslint/jsonc',
    files: [
      '**/*.jsonc',
      '**/.vscode/*.json', // VS Code settings files can use JSONC features.
      '**/jsconfig.json', // I'm not sure if `jsconfig.json` is useful here, but it's used in the same way as `tsconfig.json`. Further discussion is welcome.
      '**/jsconfig.*.json', // I'm also not sure about this. Further discussion is welcome.
      '**/tsconfig.json', // `tsconfig.json` can use JSONC features.
      '**/tsconfig.*.json', // There are `tsconfig.base.json` and `tsconfig.dist.json` files in `rewrite` repository.
    ],
    language: 'json/jsonc',
    languageOptions: {
      allowTrailingCommas: true, // Further discussion is welcome.
    },
    rules: {
      'json/no-duplicate-keys': 'error',
      'json/no-empty-keys': 'error',
      'json/no-unnormalized-keys': 'error',
      'json/no-unsafe-values': 'error',
      'json/top-level-interop': 'error',
    },
  },

  // JSON5
  {
    name: 'eslint-config-eslint/json5',
    files: ['**/*.json5'],
    language: 'json/json5',
    rules: {
      'json/no-duplicate-keys': 'error',
      'json/no-empty-keys': 'error',
      'json/no-unnormalized-keys': 'error',
      'json/no-unsafe-values': 'error',
      'json/top-level-interop': 'error',
    },
  },
];
```

**Comment by meganprice661-svg** [comment]
This looks solid. Starting with the recommended rules for JSON and Markdown
makes sense, and we can iteratively enable more rules later. Including CSS
for dogfooding also seems reasonable.

**Comment by fasttime** [comment]
> (Since Markdown has many more rulesets, I first looked into the CSS and JSON rulesets and configurations. I'll update the Markdown rulesets soon.)

I think it's fine to discuss the exact rules and rule options during the implementation, so we can also check them against our own usage.

Just one question: are you planning to start only with the ESLint plugins for the non-JavaScript languages, i.e. only `@eslint/json`, `@eslint/css` and `@eslint/markdown`? (`eslint-config-eslint` uses many third-party plugins, but only for JavaScript so far).

> Only the `code-explorer` repository is currently using a vanilla CSS file alongside Tailwind CSS, so I think CSS rules could be applied there. Since it's just one repository, I don't have a strong opinion. But, I think it's probably worth adding so we can dogfood it.
> 
>     * https://github.com/eslint/code-explorer/blob/main/src/app.css
> 
>     * https://github.com/eslint/code-explorer/blob/main/src/codemirror-themes.css

I see, that's a good point. I'm in favor of adding CSS to the `eslint-config-eslint` languages and enabling CSS file linting in `code-explorer`. That would add an additional layer of testing to `@eslint/css`.

> ### Configurations
> 
> The draft configuration is as follows. Further discussion is welcome.
> 
> ```js
> const json = require('@eslint/json').default;
> 
> // ...
> 
> const jsonConfigs = [
>   {
>     plugins: {
>       json,
>     },
>   },
> 
>   // JSON
>   {
>     name: 'eslint-config-eslint/json',
>     files: ['**/*.json', ".c8rc"],
>     ignores: [
>       '**/package-lock.json', // Ignored because lock files are usually auto generated.
>       '**/.vscode/*.json', // VS Code settings files can use JSONC features.
>       '**/jsconfig.json', // I'm not sure if `jsconfig.json` is useful here, but it's used in the same way as `tsconfig.json`. Further discussion is welcome.
>       '**/jsconfig.*.json', // I'm also not sure about this. Further discussion is welcome.
>       '**/tsconfig.json', // `tsconfig.json` can use JSONC features.
>       '**/tsconfig.*.json', // There are `tsconfig.base.json` and `tsconfig.dist.json` files in `rewrite` repository.
>     ],
>     language: 'json/json',
>     rules: {
>       'json/no-duplicate-keys': 'error',
>       'json/no-empty-keys': 'error',
>       'json/no-unnormalized-keys': 'error',
>       'json/no-unsafe-values': 'error',
>       'json/top-level-interop': 'error',
>     },
>   },
> 
>   // JSONC
>   {
>     name: 'eslint-config-eslint/jsonc',
>     files: [
>       '**/*.jsonc',
>       '**/.vscode/*.json', // VS Code settings files can use JSONC features.
>       '**/jsconfig.json', // I'm not sure if `jsconfig.json` is useful here, but it's used in the same way as `tsconfig.json`. Further discussion is welcome.
>       '**/jsconfig.*.json', // I'm also not sure about this. Further discussion is welcome.
>       '**/tsconfig.json', // `tsconfig.json` can use JSONC features.
>       '**/tsconfig.*.json', // There are `tsconfig.base.json` and `tsconfig.dist.json` files in `rewrite` repository.
>     ],
>     language: 'json/jsonc',
>     languageOptions: {
>       allowTrailingCommas: true, // Further discussion is welcome.
>     },
>     rules: {
>       'json/no-duplicate-keys': 'error',
>       'json/no-empty-keys': 'error',
>       'json/no-unnormalized-keys': 'error',
>       'json/no-unsafe-values': 'error',
>       'json/top-level-interop': 'error',
>     },
>   },
> 
>   // JSON5
>   {
>     name: 'eslint-config-eslint/json5',
>     files: ['**/*.json5'],
>     language: 'json/json5',
>     rules: {
>       'json/no-duplicate-keys': 'error',
>       'json/no-empty-keys': 'error',
>       'json/no-unnormalized-keys': 'error',
>       'json/no-unsafe-values': 'error',
>       'json/top-level-interop': 'error',
>     },
>   },
> ];
> ```

I'm unsure if we want to specify which `files` should be linted in each config. The JavaScript configs only filter files by extension:

https://github.com/eslint/eslint/blob/129882d2fdb4e7f597ed78eeadd86377f3d6b078/packages/eslint-config-eslint/index.js#L11-L18

We could also include the `recommended` configs for each plugin and simply add or remove rules as required.

Thoughts about this proposal @eslint/eslint-team?

**Comment by meganprice661-svg** [comment]
Good points.

**Comment by nzakas** [comment]
> We could also include the `recommended` configs for each plugin and simply add or remove rules as required.
> 
> Thoughts about this proposal [@eslint/eslint-team](https://github.com/orgs/eslint/teams/eslint-team)?

That seems like a fine way to get started.

**Comment by snitin315** [comment]
Yes, +1 to start with adding the default configs for each - CSS, markdown & JSON. Marking as accepted.

**Comment by JoshuaKGoldberg** [comment]
> I think using the sort-keys rule only for package.json would be useful, since all fields can be sorted and it would ensure consistency across projects. I'd like to raise this for further discussion.

I'd actually suggest the opposite approach: enforce `sort-keys` for JSON files _except_ `package.json`. The community convention for `package.json` files is to go approximately with the order set in https://docs.npmjs.com/cli/configuring-npm/package-json. This shows up in a few userland tools:

* [`eslint-plugin-package-json` > `sort-collections` > Scripts Lifecycle Ordering](https://github.com/michaelfaith/eslint-plugin-package-json/blob/591b8d948d2041a18b849bb2af680b55d67827dc/docs/rules/sort-collections.md#scripts-lifecycle-ordering): intentionally matches 👇
* [`prettier-plugin-packagejson`](https://github.com/matzkoh/prettier-plugin-packagejson): which uses 👇 
* [`sort-package-json`](https://github.com/keithamus/sort-package-json)

I _personally_ prefer everything being alphabetical, but userland seems to have spoken.

**Comment by lumirlumir** [comment]
@fasttime 

> Just one question: are you planning to start only with the ESLint plugins for the non-JavaScript languages, i.e. only @eslint/json, @eslint/css and @eslint/markdown? (eslint-config-eslint uses many third-party plugins, but only for JavaScript so far).

I think it would be nice to include some third-party plugins, if that's allowed.

As far as I recall, the userland plugins that use the language plugin feature are the three plugins below:

(There might be more userland language plugins, but these are the only three I can remember.)

- [`@html-eslint/eslint-plugin`](https://html-eslint.org/): 

    I think adding this plugin for HTML linting would be useful, since it's already included in `code-explorer` and on `eslint.org`. The HTML extension is used across ESLint repositories, so this could be a viable option.

    https://github.com/eslint/code-explorer/blob/main/package.json#L45

    https://github.com/eslint/eslint.org/blob/main/eslint.config.js#L11

- [`eslint-plugin-yml`](https://github.com/ota-meshi/eslint-plugin-yml): 

    I think linting YAML files would be nice, but it's still undergoing the RFC process for further discussion. It might be better to defer it until a decision has been made. (https://github.com/eslint/rfcs/pull/144)

- [`eslint-markdown`](https://eslint-markdown.lumir.page/): 

    This plugin provides additional Markdown rules that do not overlap with `@eslint/markdown` and aims to supply stylistic rules so that, together, the two can replace `markdownlint`.

    Although the rules don't yet fully replace `markdownlint`, the intention of this issue is to do so, and since `eslint`, `generator-eslint`, `eslint.org`, and other repositories still depend on `markdownlint`, I think this would be a useful addition.

    I created and maintain this plugin, so I may be biased—please consider it if you think it's genuinely useful. (But, I can work on creating some additional rules to better fit our use case for this plugin.)

> I'm unsure if we want to specify which files should be linted in each config. The JavaScript configs only filter files by extension:

I don't have a strong opinion, and that could be another useful option, since `tsconfig.json` can remain in JSON format unless comments are needed, and most repositories usually don't use comments in `tsconfig.json`.

> We could also include the recommended configs for each plugin and simply add or remove rules as required.

Yes, I also think it would be a good starting point.

---

@JoshuaKGoldberg 

> I personally prefer everything being alphabetical, but userland seems to have spoken.

Thanks for pointing that out! After reviewing the content, I think a userland setup for `package.json` seems more appropriate, so I'll leave it open for further discussion or future adoption.

One question: has `eslint-plugin-package-json` adopted `@eslint/json` for parsing JSON content? I recall a discussion about moving to the `@eslint/json` implementation (https://github.com/eslint/json/issues/83#issuecomment-2983800385), but I'm not sure whether that was completed or is still in progress.
